File: core/src/main/java/com/egoal/darkestpixeldungeon/items/weapon/melee/CrystalsSwords.java
package com.egoal.darkestpixeldungeon.items.weapon.melee;

import com.egoal.darkestpixeldungeon.actors.Char;
import com.egoal.darkestpixeldungeon.actors.Damage;
import com.egoal.darkestpixeldungeon.actors.hero.Hero;
import com.egoal.darkestpixeldungeon.sprites.ItemSpriteSheet;
import com.watabou.utils.Random;

/**
 * Created by 93942 on 8/16/2018.
 */

public class CrystalsSwords extends MeleeWeapon {
  {
    image = ItemSpriteSheet.DPD_CRYSTALS_SWORDS;

    tier = 3;
  }

  // 16 + 3*lvl
  @Override
  public int max(int lvl) {
    return 4 * (tier + 1) +
            lvl * (tier);
  }

  @Override
  public Damage giveDamage(Hero hero, Char target) {
    Damage dmg = super.giveDamage(hero, target);

    // 20% chance to deal 2 times damage
    float c = .15f+ .35f*(1f- (float)Math.pow(.7, level()/3));
    
    if (Random.Float() < c) {
      dmg.value *= 2.f;
      dmg.addFeature(Damage.Feature.CRITCIAL);
    }

    return dmg;
  }

}

File: core/src/main/java/com/egoal/darkestpixeldungeon/items/weapon/melee/CrystalsSwords.kt
package com.egoal.darkestpixeldungeon.items.weapon.melee

import com.egoal.darkestpixeldungeon.actors.Char
import com.egoal.darkestpixeldungeon.actors.Damage
import com.egoal.darkestpixeldungeon.actors.hero.Hero
import com.egoal.darkestpixeldungeon.sprites.ItemSpriteSheet
import com.watabou.utils.Random

/**
 * Created by 93942 on 8/16/2018.
 */

class CrystalsSwords : MeleeWeapon() {
    init {
        image = ItemSpriteSheet.DPD_CRYSTALS_SWORDS

        tier = 3
    }

    // 16 + 3*lvl
    override fun max(lvl: Int): Int = 4 * (tier + 1) + lvl * tier

    override fun giveDamage(hero: Hero, target: Char): Damage {
        val dmg = super.giveDamage(hero, target)

        // chance to deal 2 times damage
        val c = .15f + .35f * (1f - Math.pow(.7, level() / 3.0).toFloat())

        if (Random.Float() < c) {
            dmg.value *= 2
            dmg.addFeature(Damage.Feature.CRITICAL)
        }

        return dmg
    }

}
